### PR TITLE
DAT-75398 [Timbermill] implement DiskHandler for Redis

### PR DESCRIPTION
**Pull request by oz-zafar**

### CODE REVIEW

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/persistence/RedisPersistenceHandler.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void reset() {
+        List<String> queues = Arrays.asList(OVERFLOWED_EVENTS_QUEUE_NAME, FAILED_BULKS_QUEUE_NAME);
+        redisService.deleteFromRedis(queues);
```
Comment (oz-zafar) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/persistence/RedisPersistenceHandler.java:152]: in this way, the mechanism won't have access to the existing bulks keys
so at some point they will just be expired

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        return ids;
+    }
+
+    public List<String> popElementsFromRedisList(String listName, int amount) {
+        lock();
+        List<String> elements = getRangeFromRedisList(listName, 0, amount - 1);
+        trimRedisList(listName, elements.size(), -1);
+        release();
+        return elements;
+    }
```
Comment (oz-zafar) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java:236]: there is no built-in multiple POP for Redis list and issuing multiple POP requests to the Redis server seems to me too expensive,  soI implemented pop by get & cut (saw recommendations in StackOverflow)

Comment (oz-zafar) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java:236]: btw
I see our default limit for the number of fetched bulks in one call to fetchAndDelete is 10
so it's not too much.. but we can (or should?) possibly increase it

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/BulkPersistentFetchJob.java [hunk 3]
```diff
@@ -1,58 +1,58 @@
...
+            KamonConstants.CURRENT_DATA_IN_DB_GAUGE.withTag("type", "failed_bulks_amount").update(persistenceHandler.failedBulksAmount());
+            Timer.Started start = KamonConstants.BULK_FETCH_JOB_LATENCY.withoutTags().start();
+            String flowId = "Failed Bulk Persistent Fetch Job - " + UUID.randomUUID().toString();
+            MDC.put("id", flowId);
+            LOG.info("Failed Bulks Persistent Fetch Job started.");
+            ElasticsearchClient es = (ElasticsearchClient) context.getJobDetail().getJobDataMap().get(CLIENT);
+            retryFailedRequests(es, persistenceHandler);
+            LOG.info("Failed Bulks Persistent Fetch Job ended.");
+            start.stop();
+        }
```
Comment (oz-zafar) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/BulkPersistentFetchJob.java:39]: I considered adding a lock here -
there is no point that each pod would fetch the same queue.
Though it's specifically to the Redis implementation so maybe it's not such a good idea

**Review by edengal — CHANGES_REQUESTED** [review]

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java (hunk 2 above)
Comment (oz-zafar) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java:238]: @edengal what about a sorted set? it has SPOP too

Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java:238]: Sure, use ZADD with a score representing the time and ZPOPMIN to get the earliest, it a bit slower but still pretty fast.
@oz-zafar

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/cache/RedisCacheHandler.java [hunk 4]
```diff
@@ -1,113 +1,48 @@
...
+        Map<String, List<String>> orphans = redisService.getFromRedis(orphanParentsIds);
 
         Map<String, List<String>> retMap = Maps.newHashMap();
         for (Map.Entry<String, List<String>> entry : orphans.entrySet()) {
             String newKey = entry.getKey().substring(ORPHAN_PREFIX.length());
             retMap.put(newKey, entry.getValue());
         }
-
+        if (retMap.size() < orphanParentsIds.size()){
+//            LOG.warn("Failed to pull some ids from Redis orphans cache. Ids: {}", Sets.difference(orphanParentsIds, retMap.keySet()));
```
Comment (edengal) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/cache/RedisCacheHandler.java:44]: If no need delete

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/cache/RedisCacheHandler.java [hunk 5]
```diff
@@ -118,115 +53,40 @@ public void pushToOrphanCache(Map<String, List<String>> orphansMap) {
...
-                        kryoPool.free(kryo);
-                    }
-
-                }
-            } catch (Exception e) {
-                LOG.error("Error getting ids from Redis " + cacheName + " cache. Ids: " + idsPartition , e);
-            }
+        Map<String, LocalTask> retMap = redisService.getFromRedis(idsList);
+        if (retMap.size() < idsList.size()){
+//            LOG.warn("Failed to pull some ids from Redis tasks cache. Ids: {}", Sets.difference(Sets.newHashSet(idsList), retMap.keySet()));
```
Comment (edengal) [inline timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/cache/RedisCacheHandler.java:65]: If no need delete

**Review by edengal — APPROVED** [review]

**Pull request merged**